File: ml-service/backtest_pipeline/grid_v2.py
"""
TURBO-BOT P#71 — Grid V2 Strategy Module
Dedicated mean-reversion grid trading for RANGING regime.

KEY DIFFERENCES from P#67/P#68 Grid:
- BYPASSES ensemble voting entirely (direct execution)
- Uses Bollinger Band boundaries as grid edges
- Tight SL/TP (0.5-0.8 ATR) for quick flips
- Only activates in RANGING (ADX < grid_adx_threshold)
- Has its own confidence calculation (no ensemble dilution)
- Separate position from directional trades

HOW IT WORKS:
1. Detect RANGING regime (ADX < 22, confirmed by low directional movement)
2. Calculate grid levels from Bollinger Bands (upper/lower ± adjustments)
3. BUY at lower grid (price near BB lower + RSI < 45)
4. SELL at upper grid (price near BB upper + RSI > 55)
5. Tight SL (0.6 ATR) + Tight TP (0.8 ATR)
6. Cooldown between grid trades (min 4 candles = 1h on 15m)
7. Max grid trades per session (prevent overtrading)

EXPECTED IMPACT:
- BTC: Major source of alpha (BTC sideways 40-60% of time)
- ETH: Secondary (good ranging periods)
- BNB: Supplement to SHORT bias
- SOL: Minimal (SOL trends more, less ranging)
"""

import numpy as np
from . import config


# ============================================================================
# GRID V2 DEFAULT PARAMETERS
# ============================================================================

GRID_V2_ADX_THRESHOLD = 20       # Max ADX for grid activation (stricter)
GRID_V2_BB_LOWER_ENTRY = 0.08    # BB%B below this → BUY zone (extreme only)
GRID_V2_BB_UPPER_ENTRY = 0.92    # BB%B above this → SELL zone (extreme only)
GRID_V2_RSI_OVERSOLD = 38        # RSI below this for BUY confirmation
GRID_V2_RSI_OVERBOUGHT = 62      # RSI above this for SELL confirmation
GRID_V2_SL_ATR = 0.70            # SL (0.7 × ATR)
GRID_V2_TP_ATR = 1.20            # TP (1.2 × ATR) — covers fees
GRID_V2_COOLDOWN = 16            # Min candles between grid trades (4h on 15m)
GRID_V2_MAX_TRADES = 25          # Max grid trades per backtest session
GRID_V2_RISK_PER_TRADE = 0.006   # 0.6% risk per grid trade
GRID_V2_MIN_BB_WIDTH = 0.008     # Min BB width (avoid dead periods)


class GridV2Strategy:
    """
    Dedicated Grid V2 mean-reversion strategy for RANGING markets.
    
    This operates INDEPENDENTLY from the ensemble pipeline.
    It generates its own signals with its own confidence calculation.
    """
    
    def __init__(self, symbol='BTCUSDT'):
        self.symbol = symbol
        self.enabled = True
        
        # State tracking
        self.last_grid_candle = -999
        self.grid_trades_count = 0
        self.grid_wins = 0
        self.grid_losses = 0
        self.grid_pnl = 0
        self.total_grid_fees = 0
        self.consecutive_grid_wins = 0
        self.max_consecutive_wins = 0
        
        # Per-pair configurable parameters
        self.adx_threshold = GRID_V2_ADX_THRESHOLD
        self.bb_lower_entry = GRID_V2_BB_LOWER_ENTRY
        self.bb_upper_entry = GRID_V2_BB_UPPER_ENTRY
        self.rsi_oversold = GRID_V2_RSI_OVERSOLD
        self.rsi_overbought = GRID_V2_RSI_OVERBOUGHT
        self.sl_atr = GRID_V2_SL_ATR
        self.tp_atr = GRID_V2_TP_ATR
        self.cooldown = GRID_V2_COOLDOWN
        self.max_trades = GRID_V2_MAX_TRADES
        self.risk_per_trade = GRID_V2_RISK_PER_TRADE
        self.min_bb_width = GRID_V2_MIN_BB_WIDTH
        
        # P#72: Adaptive base values (saved for reset after streak)
        self._base_cooldown = GRID_V2_COOLDOWN
        self._base_bb_lower = GRID_V2_BB_LOWER_ENTRY
        self._base_bb_upper = GRID_V2_BB_UPPER_ENTRY
        self._consecutive_losses = 0
        
    def configure(self, **kwargs):
        """Apply per-pair configuration overrides."""
        mapping = {
            'GRID_V2_ADX_THRESHOLD': 'adx_threshold',
            'GRID_V2_BB_LOWER_ENTRY': 'bb_lower_entry',
            'GRID_V2_BB_UPPER_ENTRY': 'bb_upper_entry',
            'GRID_V2_RSI_OVERSOLD': 'rsi_oversold',
            'GRID_V2_RSI_OVERBOUGHT': 'rsi_overbought',
            'GRID_V2_SL_ATR': 'sl_atr',
            'GRID_V2_TP_ATR': 'tp_atr',
            'GRID_V2_COOLDOWN': 'cooldown',
            'GRID_V2_MAX_TRADES': 'max_trades',
            'GRID_V2_RISK_PER_TRADE': 'risk_per_trade',
            'GRID_V2_MIN_BB_WIDTH': 'min_bb_width',
        }
        for cfg_key, attr_name in mapping.items():
            if cfg_key in kwargs:
                setattr(self, attr_name, kwargs[cfg_key])
        # P#72: Save base values for adaptive reset
        self._base_cooldown = self.cooldown
        self._base_bb_lower = self.bb_lower_entry
        self._base_bb_upper = self.bb_upper_entry
    
    def evaluate(self, row, history, regime, candle_idx, has_position=False):
        """
        Evaluate whether a grid trade should be taken.
        
        This BYPASSES the ensemble — generates independent signals.
        Only fires in RANGING regime with mean-reversion conditions.
        
        Args:
            row: current candle
            history: recent history
            regime: current market regime
            candle_idx: current candle index
            has_position: whether there's already an open position
            
        Returns:
            dict: {
                'signal': 'BUY'|'SELL'|None,
                'confidence': float,
                'is_grid': True,
                'sl_atr': float,
                'tp_atr': float,
                'risk_per_trade': float,
                'reason': str,
            } or None if no signal
        """
        if not self.enabled:
            return None
            
        # Don't generate grid signal if there's already a position
        if has_position:
            return None
            
        # Max trades limit
        if self.grid_trades_count >= self.max_trades:
            return None
            
        # Cooldown check
        if (candle_idx - self.last_grid_candle) < self.cooldown:
            return None
        
        # === REGIME GATE: Only RANGING ===
        adx = row.get('adx', 30)
        if adx > self.adx_threshold:
            return None
        if regime not in ('RANGING', 'HIGH_VOLATILITY'):
            # Allow HV if ADX is low (choppy HV = still grid-tradeable)
            if regime != 'HIGH_VOLATILITY' or adx > 18:
                return None
        
        # === BOLLINGER BAND ANALYSIS ===
        bb_pctb = row.get('bb_pctb', 0.5)
        close = row['close']
        bb_upper = row.get('bb_upper', close * 1.02)
        bb_lower = row.get('bb_lower', close * 0.98)
        
        # Check BB width (avoid flat/dead periods)
        bb_width = (bb_upper - bb_lower) / close if close > 0 else 0
        if bb_width < self.min_bb_width:
            return None
        
        # === RSI CONFIRMATION ===
        rsi = row.get('rsi_14', 50)
        
        # === SIGNAL GENERATION ===
        signal = None
        confidence = 0
        reason = ''
        
        # BUY zone: price near BB lower + RSI oversold
        if bb_pctb < self.bb_lower_entry and rsi < self.rsi_oversold:
            signal = 'BUY'
            # Confidence based on how deep in the zone
            depth = (self.bb_lower_entry - bb_pctb) / self.bb_lower_entry
            rsi_strength = (self.rsi_oversold - rsi) / self.rsi_oversold
            confidence = 0.30 + depth * 0.30 + rsi_strength * 0.20
            reason = f'Grid BUY: BB%B={bb_pctb:.2f} RSI={rsi:.0f} ADX={adx:.0f}'
            
        # SELL zone: price near BB upper + RSI overbought  
        elif bb_pctb > self.bb_upper_entry and rsi > self.rsi_overbought:
            signal = 'SELL'
            depth = (bb_pctb - self.bb_upper_entry) / (1 - self.bb_upper_entry)
            rsi_strength = (rsi - self.rsi_overbought) / (100 - self.rsi_overbought)
            confidence = 0.30 + depth * 0.30 + rsi_strength * 0.20
            reason = f'Grid SELL: BB%B={bb_pctb:.2f} RSI={rsi:.0f} ADX={adx:.0f}'
        
        if signal is None:
            return None
        
        # Clamp confidence
        confidence = np.clip(confidence, 0.20, 0.75)
        
        # Volume confirmation bonus
        vol_ratio = row.get('volume_ratio', 1.0)
        if vol_ratio > 1.3:
            confidence += 0.05
            reason += f' Vol={vol_ratio:.1f}x'
        
        # Streak bonus: consecutive grid wins → slightly higher confidence
        if self.consecutive_grid_wins >= 3:
            confidence += 0.03
        
        return {
            'signal': signal,
            'confidence': float(np.clip(confidence, 0.20, 0.80)),
            'is_grid': True,
            'sl_atr': self.sl_atr,
            'tp_atr': self.tp_atr,
            'risk_per_trade': self.risk_per_trade,
            'reason': reason,
        }
    
    def record_grid_trade(self, pnl, fees=0):
        """Record a completed grid trade for tracking."""
        self.grid_trades_count += 1
        self.grid_pnl += pnl
        self.total_grid_fees += fees
        
        if pnl > 0:
            self.grid_wins += 1
            self.consecutive_grid_wins += 1
            self._consecutive_losses = 0
            self.max_consecutive_wins = max(
                self.max_consecutive_wins, self.consecutive_grid_wins)
            # P#72 Adaptive: relax cooldown after 3 consecutive wins
            if self.consecutive_grid_wins >= 3:
                self.cooldown = max(self._base_cooldown - 4, 8)
        else:
            self.grid_losses += 1
            self.consecutive_grid_wins = 0
            self._consecutive_losses += 1
            # P#72 Adaptive: tighten cooldown after 2 consecutive losses
            if self._consecutive_losses >= 2:
                self.cooldown = min(self._base_cooldown + 8, 48)
                # Also tighten entry zones temporarily
                self.bb_lower_entry = max(self._base_bb_lower - 0.02, 0.03)
                self.bb_upper_entry = min(self._base_bb_upper + 0.02, 0.97)
            # Reset to base after single loss (not streak)
            if self._consecutive_losses == 1:
                self.cooldown = self._base_cooldown
                self.bb_lower_entry = self._base_bb_lower
                self.bb_upper_entry = self._base_bb_upper
    
    def mark_entry(self, candle_idx):
        """Mark that a grid trade was opened at this candle."""
        self.last_grid_candle = candle_idx
    
    def get_stats(self):
        """Get grid V2 statistics."""
        win_rate = (self.grid_wins / max(self.grid_trades_count, 1)) * 100
        pf = 0
        if self.grid_losses > 0 and self.grid_wins > 0:
            avg_win = self.grid_pnl / max(self.grid_wins, 1) if self.grid_pnl > 0 else 0
            # Approximate PF from win/loss ratio and win rate
            pf = (win_rate / 100 * abs(self.grid_pnl)) / max(abs(self.grid_pnl - self.grid_pnl), 0.01)
        
        return {
            'enabled': self.enabled,
            'total_trades': self.grid_trades_count,
            'wins': self.grid_wins,
            'losses': self.grid_losses,
            'win_rate': round(win_rate, 1),
            'net_pnl': round(self.grid_pnl, 4),
            'total_fees': round(self.total_grid_fees, 4),
            'max_consecutive_wins': self.max_consecutive_wins,
            'adx_threshold': self.adx_threshold,
        }
    
    def reset(self):
        """Reset for fresh run."""
        self.last_grid_candle = -999
        self.grid_trades_count = 0
        self.grid_wins = 0
        self.grid_losses = 0
        self.grid_pnl = 0
        self.total_grid_fees = 0
        self.consecutive_grid_wins = 0
        self.max_consecutive_wins = 0
        self._consecutive_losses = 0
        # P#72: Reset adaptive params to base
        self.cooldown = self._base_cooldown
        self.bb_lower_entry = self._base_bb_lower
        self.bb_upper_entry = self._base_bb_upper

File: trading-bot/src/strategies/grid-v2.js
'use strict';
/**
 * @module GridV2Strategy
 * @description P#171: Live JS implementation of Grid V2 mean-reversion strategy.
 * Ported from ml-service/backtest_pipeline/grid_v2.py (P#71/P#72).
 *
 * RANGING-only (ADX < threshold), BB%B + RSI mean-reversion.
 * BYPASSES ensemble voting — operates independently.
 *
 * TARGET PAIRS: BNBUSDT ($3800), XRPUSDT ($1000)
 */

// ============================================================================
// GRID V2 PARAMETERS (P#211: updated from Python backtest, calibrated)
// ============================================================================

const GRID_ADX_THRESHOLD = 20;        // Max ADX for grid activation
const GRID_BB_LOWER_ENTRY = 0.12;     // P#200c: Relaxed from 0.08 — more entries
const GRID_BB_UPPER_ENTRY = 0.88;     // P#200c: Relaxed from 0.92
const GRID_RSI_OVERSOLD = 36;         // RSI below → BUY confirmation
const GRID_RSI_OVERBOUGHT = 64;       // RSI above → SELL confirmation
const GRID_SL_ATR = 0.70;             // SL 0.7× ATR (tight)
const GRID_TP_ATR = 1.50;             // P#211b: widened from 1.20 — fee economics require bigger TP
const GRID_COOLDOWN_MS = 5 * 3600000; // 5h between grid trades
const GRID_MAX_TRADES_DAY = 8;        // max grid trades per day
const GRID_RISK_PER_TRADE = 0.006;    // 0.6% risk per grid trade (overridden per-pair in bot.js)
const GRID_MIN_BB_WIDTH = 0.008;      // min BB width (avoid dead periods)

class GridV2Strategy {
    /**
     * @param {string} symbol — e.g. 'BNBUSDT'
     * @param {Object} [opts] — override defaults
     */
    constructor(symbol, opts = {}) {
        this.symbol = symbol;
        this.enabled = true;

        // Configurable parameters
        this.adxThreshold = opts.adxThreshold || GRID_ADX_THRESHOLD;
        this.bbLowerEntry = opts.bbLowerEntry || GRID_BB_LOWER_ENTRY;
        this.bbUpperEntry = opts.bbUpperEntry || GRID_BB_UPPER_ENTRY;
        this.rsiOversold = opts.rsiOversold || GRID_RSI_OVERSOLD;
        this.rsiOverbought = opts.rsiOverbought || GRID_RSI_OVERBOUGHT;
        this.slAtr = opts.slAtr || GRID_SL_ATR;
        this.tpAtr = opts.tpAtr || GRID_TP_ATR;
        this.cooldownMs = opts.cooldownMs || GRID_COOLDOWN_MS;
        this.maxTradesDay = opts.maxTradesDay || GRID_MAX_TRADES_DAY;
        this.riskPerTrade = opts.riskPerTrade || GRID_RISK_PER_TRADE;
        this.minBbWidth = opts.minBbWidth || GRID_MIN_BB_WIDTH;

        // State
        this._lastGridTradeTime = 0;
        this._gridTradesToday = 0;
        this._dayResetTime = Date.now();
        this._consecutiveLosses = 0;
        this._consecutiveWins = 0;

        // Tracking
        this.gridWins = 0;
        this.gridLosses = 0;
        this.gridPnL = 0;
        this.totalGridFees = 0;
        this.totalGridTrades = 0;
        this.maxConsecutiveWins = 0;

        // P#72: Adaptive base values
        this._baseCooldownMs = this.cooldownMs;
        this._baseBbLower = this.bbLowerEntry;
        this._baseBbUpper = this.bbUpperEntry;
    }

    /**
     * Evaluate whether a grid trade should be taken.
     * BYPASSES ensemble — generates independent signals.
     *
     * @param {Object} params
     * @param {number} params.currentPrice
     * @param {Object} params.indicators — { rsi, adx, bb_pctb, bb_upper, bb_lower, atr, volume_ratio }
     * @param {string} params.regime — current market regime
     * @param {boolean} params.hasPosition — whether bot already has a position
     * @returns {Object|null} — { action, confidence, sl, tp, riskPerTrade, reason, isGrid } or null
     */
    evaluate(params) {
        const { currentPrice, indicators, regime, hasPosition } = params;

        if (!this.enabled) return null;
        if (hasPosition) return null;

        // Daily trade limit
        if (Date.now() - this._dayResetTime > 86400000) {
            this._gridTradesToday = 0;
            this._dayResetTime = Date.now();
        }
        if (this._gridTradesToday >= this.maxTradesDay) return null;

        // Cooldown check
        if (Date.now() - this._lastGridTradeTime < this.cooldownMs) return null;

        // === REGIME GATE: ADX-based (P#216: removed strict RANGING requirement) ===
        // Grid works on ANY regime where ADX is low (mean-reverting conditions).
        // Previously required NeuralAI regime='RANGING' which often returned 'UNKNOWN'.
        const adx = indicators.adx || 30;
        if (adx > this.adxThreshold) return null;
        // Only block during strong trending (TRENDING_UP/DOWN with high ADX confirmed)
        if ((regime === 'TRENDING_UP' || regime === 'TRENDING_DOWN') && adx > 15) return null;

        // === BOLLINGER BAND ANALYSIS ===
        const bbPctb = indicators.bb_pctb;
        if (typeof bbPctb !== 'number') return null;

        const bbUpper = indicators.bb_upper || currentPrice * 1.02;
        const bbLower = indicators.bb_lower || currentPrice * 0.98;

        // BB width check — avoid flat/dead periods
        const bbWidth = currentPrice > 0 ? (bbUpper - bbLower) / currentPrice : 0;
        if (bbWidth < this.minBbWidth) return null;

        // === RSI CONFIRMATION ===
        const rsi = indicators.rsi || 50;

        // === SIGNAL GENERATION ===
        let action = null;
        let confidence = 0;
        let reason = '';
        const atr = indicators.atr || currentPrice * 0.01;

        // BUY zone: price near BB lower + RSI oversold
        if (bbPctb < this.bbLowerEntry && rsi < this.rsiOversold) {
            action = 'BUY';
            const depth = (this.bbLowerEntry - bbPctb) / this.bbLowerEntry;
            const rsiStrength = (this.rsiOversold - rsi) / this.rsiOversold;
            confidence = 0.30 + depth * 0.30 + rsiStrength * 0.20;
            reason = 'Grid BUY: BB%B=' + bbPctb.toFixed(2) + ' RSI=' + rsi.toFixed(0) + ' ADX=' + adx.toFixed(0);
        }
        // SELL zone: price near BB upper + RSI overbought
        else if (bbPctb > this.bbUpperEntry && rsi > this.rsiOverbought) {
            action = 'SELL';
            const depth = (bbPctb - this.bbUpperEntry) / (1 - this.bbUpperEntry);
            const rsiStrength = (rsi - this.rsiOverbought) / (100 - this.rsiOverbought);
            confidence = 0.30 + depth * 0.30 + rsiStrength * 0.20;
            reason = 'Grid SELL: BB%B=' + bbPctb.toFixed(2) + ' RSI=' + rsi.toFixed(0) + ' ADX=' + adx.toFixed(0);
        }

        if (!action) return null;

        // Clamp confidence
        confidence = Math.max(0.20, Math.min(0.75, confidence));

        // Volume confirmation bonus
        const volRatio = indicators.volume_ratio || 1.0;
        if (volRatio > 1.3) {
            confidence += 0.05;
            reason += ' Vol=' + volRatio.toFixed(1) + 'x';
        }

        // Streak bonus
        if (this._consecutiveWins >= 3) confidence += 0.03;

        confidence = Math.max(0.20, Math.min(0.80, confidence));

        // Calculate SL/TP from ATR
        const sl = action === 'BUY'
            ? currentPrice - this.slAtr * atr
            : currentPrice + this.slAtr * atr;
        const tp = action === 'BUY'
            ? currentPrice + this.tpAtr * atr
            : Math.max(0.0000001, currentPrice - this.tpAtr * atr);

        return {
            action,
            confidence,
            sl,
            tp,
            riskPerTrade: this.riskPerTrade,
            reason,
            isGrid: true,
            strategy: 'GridV2',
        };
    }

    /**
     * Record result of a completed grid trade.
     * @param {number} pnl — profit/loss
     * @param {number} [fees] — fees paid
     */
    recordTradeResult(pnl, fees = 0) {
        this.totalGridTrades++;
        this.gridPnL += pnl;
        this.totalGridFees += fees;
        this._gridTradesToday++;
        this._lastGridTradeTime = Date.now();

        if (pnl > 0) {
            this.gridWins++;
            this._consecutiveWins++;
            this._consecutiveLosses = 0;
            this.maxConsecutiveWins = Math.max(this.maxConsecutiveWins, this._consecutiveWins);

            // P#72 Adaptive: relax cooldown after 3 consecutive wins
            if (this._consecutiveWins >= 3) {
                this.cooldownMs = Math.max(this._baseCooldownMs - 3600000, 2 * 3600000); // min 2h
            }
        } else {
            this.gridLosses++;
            this._consecutiveWins = 0;
            this._consecutiveLosses++;

            // P#72 Adaptive: tighten after 2+ consecutive losses
            // P#237: FIXED — after losses, TIGHTEN entries (higher bbLower = need MORE oversold, lower bbUpper = need MORE overbought)
            if (this._consecutiveLosses >= 2) {
                this.cooldownMs = Math.min(this._baseCooldownMs + 2 * 3600000, 12 * 3600000); // max 12h
                this.bbLowerEntry = Math.min(this._baseBbLower + 0.02, 0.25);
                this.bbUpperEntry = Math.max(this._baseBbUpper - 0.02, 0.75);
            }
            // Reset after single loss
            if (this._consecutiveLosses === 1) {
                this.cooldownMs = this._baseCooldownMs;
                this.bbLowerEntry = this._baseBbLower;
                this.bbUpperEntry = this._baseBbUpper;
            }
        }
    }

    getStats() {
        const winRate = this.totalGridTrades > 0
            ? (this.gridWins / this.totalGridTrades * 100) : 0;

        return {
            enabled: this.enabled,
            totalTrades: this.totalGridTrades,
            wins: this.gridWins,
            losses: this.gridLosses,
            winRate: Math.round(winRate * 10) / 10,
            netPnL: Math.round(this.gridPnL * 10000) / 10000,
            totalFees: Math.round(this.totalGridFees * 10000) / 10000,
            maxConsecutiveWins: this.maxConsecutiveWins,
            adxThreshold: this.adxThreshold,
            currentlyOpen: false, // grid doesn't track its own position (bot does)
        };
    }

    exportState() {
        return {
            gridV2: {
                _lastGridTradeTime: this._lastGridTradeTime,
                _gridTradesToday: this._gridTradesToday,
                _dayResetTime: this._dayResetTime,
                _consecutiveLosses: this._consecutiveLosses,
                _consecutiveWins: this._consecutiveWins,
                gridWins: this.gridWins,
                gridLosses: this.gridLosses,
                gridPnL: this.gridPnL,
                totalGridFees: this.totalGridFees,
                totalGridTrades: this.totalGridTrades,
                maxConsecutiveWins: this.maxConsecutiveWins,
                cooldownMs: this.cooldownMs,
                bbLowerEntry: this.bbLowerEntry,
                bbUpperEntry: this.bbUpperEntry,
            },
        };
    }

    restoreState(data) {
        if (!data || !data.gridV2) return;
        const s = data.gridV2;
        this._lastGridTradeTime = s._lastGridTradeTime || 0;
        this._gridTradesToday = s._gridTradesToday || 0;
        this._dayResetTime = s._dayResetTime || Date.now();
        this._consecutiveLosses = s._consecutiveLosses || 0;
        this._consecutiveWins = s._consecutiveWins || 0;
        this.gridWins = s.gridWins || 0;
        this.gridLosses = s.gridLosses || 0;
        this.gridPnL = s.gridPnL || 0;
        this.totalGridFees = s.totalGridFees || 0;
        this.totalGridTrades = s.totalGridTrades || 0;
        this.maxConsecutiveWins = s.maxConsecutiveWins || 0;
        if (s.cooldownMs) this.cooldownMs = s.cooldownMs;
        if (s.bbLowerEntry) this.bbLowerEntry = s.bbLowerEntry;
        if (s.bbUpperEntry) this.bbUpperEntry = s.bbUpperEntry;
    }
}

module.exports = { GridV2Strategy };
